pkg/controller/persistentvolume/persistentvolume_controller.go

```diff
@@ -0,0 +1,1438 @@
+	}
+
+	if _, exists := ctrl.stopChannels["claims"]; !exists {
+		ctrl.stopChannels["claims"] = make(chan struct{})
+		go ctrl.claimController.Run(ctrl.stopChannels["claims"])
+	}
+}
+
+// Stop gracefully shuts down this controller
+func (ctrl *PersistentVolumeController) Stop() {
+	glog.V(5).Infof("Stopping PersistentVolumeController")
+	for name, stopChan := range ctrl.stopChannels {
+		close(stopChan)
+		delete(ctrl.stopChannels, name)
+	}
+}
+
+// isFullySynced returns true, if both volume and claim caches are fully loaded
+// after startup.
+// We do not want to process events with not fully loaded caches - e.g. we might
+// recycle/delete PVs that don't have corresponding claim in the cache yet.
+func (ctrl *PersistentVolumeController) isFullySynced() bool {
+	return ctrl.volumeController.HasSynced() && ctrl.claimController.HasSynced()
+}
+
+func (ctrl *PersistentVolumeController) hasAnnotation(obj api.ObjectMeta, ann string) bool {
+	_, found := obj.Annotations[ann]
+	return found
+}
+
+func (ctrl *PersistentVolumeController) setAnnotation(obj *api.ObjectMeta, ann string, value string) {
+	if obj.Annotations == nil {
+		obj.Annotations = make(map[string]string)
+	}
+	obj.Annotations[ann] = value
+}
+
+func (ctrl *PersistentVolumeController) updateClaimPhase(claim *api.PersistentVolumeClaim, phase api.PersistentVolumeClaimPhase) (*api.PersistentVolumeClaim, error) {
+	glog.V(5).Infof("Updating PersistentVolumeClaim[%s]: set phase %s", claimToClaimKey(claim), phase)
+	if claim.Status.Phase == phase {
+		// Nothing to do.
+		glog.V(5).Infof("Updating PersistentVolumeClaim[%s]: phase %s already set", claimToClaimKey(claim), phase)
+		return claim, nil
+	}
+
+	clone, err := conversion.NewCloner().DeepCopy(claim)
+	if err != nil {
+		return nil, fmt.Errorf("Error cloning claim: %v", err)
+	}
+	claimClone, ok := clone.(*api.PersistentVolumeClaim)
+	if !ok {
+		return nil, fmt.Errorf("Unexpected claim cast error : %v", claimClone)
+	}
+
+	claimClone.Status.Phase = phase
+	newClaim, err := ctrl.kubeClient.Core().PersistentVolumeClaims(claimClone.Namespace).UpdateStatus(claimClone)
+	if err != nil {
+		glog.V(5).Infof("Updating PersistentVolumeClaim[%s]: set phase %s failed: %v", claimToClaimKey(claim), phase, err)
+		return newClaim, err
+	}
+	glog.V(5).Infof("Updating PersistentVolumeClaim[%s]: phase %s set", claimToClaimKey(claim), phase)
+	return newClaim, nil
+}
+
+func (ctrl *PersistentVolumeController) updateClaimPhaseWithEvent(claim *api.PersistentVolumeClaim, phase api.PersistentVolumeClaimPhase, eventtype, reason, message string) (*api.PersistentVolumeClaim, error) {
+	glog.V(5).Infof("Updating updateClaimPhaseWithEvent[%s]: set phase %s", claimToClaimKey(claim), phase)
+	if claim.Status.Phase == phase {
+		// Nothing to do.
+		glog.V(5).Infof("Updating updateClaimPhaseWithEvent[%s]: phase %s already set", claimToClaimKey(claim), phase)
+		return claim, nil
+	}
+
+	// Emit the event only when the status change happens, not everytime
```

Comment:
And it broke the tests... Mock event recorder does not have any locks in it and golang race detector catches the controller emitting new event and test framework checking the event as a race. I need to add graceful shutdown of the controller to the tests...